### PR TITLE
feat: Optimizely Factory Revised

### PR DESCRIPTION
**Pull request by oakbani**
## Summary
- Add convenience methods to Optimizely Factory. set_polling_interval and set_blocking_timeout
- Refactor Factory to assign same logger, notification center instances to both config manager and actual client instance

## Test plan
Revised test cases to assert our expected custom components with values we set.

### CODE REVIEW

File: lib/optimizely/config_manager/http_project_config_manager.rb [hunk 1]
```diff
@@ -30,7 +30,7 @@ module Optimizely
   class HTTPProjectConfigManager < ProjectConfigManager
     # Config manager that polls for the datafile and updated ProjectConfig based on an update interval.
 
-    attr_reader :stopped
+    attr_reader :blocking_timeout, :error_handler, :logger, :notification_center, :polling_interval, :stopped
```
Comment (msohailhussain) [inline lib/optimizely/config_manager/http_project_config_manager.rb]: I see no issue to make reader attribute, but plz specify reason if those are helping in unit testing.

File: lib/optimizely/config_manager/http_project_config_manager.rb [hunk 2]
```diff
@@ -197,7 +197,7 @@ def set_config(config)
         "Old revision number: #{previous_revision}. New revision number: #{@config.revision}.")
     end
 
-    def polling_interval(polling_interval)
+    def set_polling_interval(polling_interval)
```
Comment (msohailhussain) [inline lib/optimizely/config_manager/http_project_config_manager.rb]: we can't really rename it.

File: lib/optimizely/config_manager/http_project_config_manager.rb [hunk 3]
```diff
@@ -234,7 +234,7 @@ def polling_interval(polling_interval)
       @polling_interval = polling_interval
     end
 
-    def blocking_timeout(blocking_timeout)
+    def set_blocking_timeout(blocking_timeout)
```
Comment (msohailhussain) [inline lib/optimizely/config_manager/http_project_config_manager.rb]: same here.

File: lib/optimizely/optimizely_factory.rb [hunk 4]
```diff
@@ -19,14 +19,16 @@
 require 'optimizely'
 require 'optimizely/event_dispatcher'
 require 'optimizely/event/batch_event_processor'
+require 'optimizely/error_handler'
```
Comment (msohailhussain) [inline lib/optimizely/optimizely_factory.rb]: alphabetize

File: lib/optimizely/optimizely_factory.rb [hunk 5]
```diff
@@ -19,14 +19,16 @@
...
+
 module Optimizely
   class OptimizelyFactory
-    attr_reader :max_event_batch_size, :max_event_flush_interval
-
     # Convenience method for setting the maximum number of events contained within a batch.
     # @param batch_size Integer - Sets size of EventQueue.
     # @param logger - Optional LoggerInterface Provides a log method to log messages.
-    def self.max_event_batch_size(batch_size, logger)
+    def self.set_max_event_batch_size(batch_size, logger = NoOpLogger.new)
```
Comment (msohailhussain) [inline lib/optimizely/optimizely_factory.rb]: same here, can't rename.

File: spec/optimizely_factory_spec.rb [hunk 6]
```diff
@@ -30,22 +30,37 @@
...
+      .with(
+        headers: {
+          'Content-Type' => 'application/json'
+        }
       )
+      .to_return(status: 200, body: '', headers: {})
+  end
 
+  describe '.default_instance' do
+    it 'should create http config manager when sdk is given' do
```
Comment (msohailhussain) [inline spec/optimizely_factory_spec.rb]: `sdk` is given?

File: spec/optimizely_factory_spec.rb [hunk 7]
```diff
@@ -30,22 +30,37 @@
...
       optimizely_instance = Optimizely::OptimizelyFactory.default_instance('sdk_key', datafile)
+      expect(optimizely_instance.config_manager). to be_instance_of(Optimizely::HTTPProjectConfigManager)
+    end
 
-      expect(optimizely_instance.config_manager). to eq(http_project_config_manager)
+    it 'should create http config manager when polling interval and blocking timeout are set' do
```
Comment (msohailhussain) [inline spec/optimizely_factory_spec.rb]: Please add in a comment, verifying polling interval and blocking timeouts are returned from config_manager when set using optimizely factory.

File: spec/optimizely_factory_spec.rb [hunk 8]
```diff
@@ -57,98 +72,98 @@ class CustomConfigManager
...
-        sdk_key: 'sdk_key',
-        datafile: datafile,
-        logger: spy_logger,
-        error_handler: error_handler,
-        skip_json_validation: false,
-        notification_center: notification_center
-      )
-
+    it 'should take http config manager when sdk key, polling interval, blocking timeout are given' do
+      Optimizely::OptimizelyFactory.set_polling_interval(50)
```
Comment (msohailhussain) [inline spec/optimizely_factory_spec.rb]: Do we really need this case, it's already defined. 
https://github.com/optimizely/ruby-sdk/pull/204/files#diff-a0f73069c4d39dc48b54a5beba886129R50

File: lib/optimizely/config_manager/http_project_config_manager.rb (hunk 2 above)
Comment (oakbani) [reply lib/optimizely/config_manager/http_project_config_manager.rb]: I don't see any issue in renaming. These are private methods of the class.

File: spec/optimizely_factory_spec.rb (hunk 8 above)
Comment (oakbani) [reply spec/optimizely_factory_spec.rb]: Yes, this checks custom_instance convenience method. While the other one verifies default_instance method

File: lib/optimizely/config_manager/http_project_config_manager.rb (hunk 1 above)
Comment (oakbani) [reply lib/optimizely/config_manager/http_project_config_manager.rb]: Yes, helps with unit testing.

**Pull request merged**